### ISSUE TITLE
strlen() expects parameter 1 to be string, array given | magento/module-catalog-search

### ISSUE THREAD

**Issue by Morgy93** [magento/magento2#33589]
### Preconditions (*)

- Magento 2.4.2
- PHP 7.4

### Steps to reproduce (*)

1. Go to `/catalogsearch/advanced/result/?price[from][]=1&price[to]=1`
2. On trying to access above URL from front-end, "There has been an error processing your request" error message is getting displayed on UI:
http://magento2.4.local/catalogsearch/advanced/result/?price[from][]=1&price[to]=1 
![image](https://user-images.githubusercontent.com/60198231/127654716-9420ae56-2ef8-494f-9107-006ab679bd45.png)

### Expected result (*)

It should not throw an unhandled error.

### Actual result (*)

```
Exception #0 (Exception): Warning: strlen() expects parameter 1 to be string, array given in vendor/magento/module-catalog-search/Model/Advanced.php on line 380

#1 strlen() called at [vendor/magento/module-catalog-search/Model/Advanced.php:380]
#2 Magento\CatalogSearch\Model\Advanced->getPreparedSearchCriteria() called at [vendor/magento/module-catalog-search/Model/Advanced.php:206]
#3 Magento\CatalogSearch\Model\Advanced->addFilters() called at [vendor/magento/module-catalog-search/Controller/Advanced/Result.php:62]
#4 Magento\CatalogSearch\Controller\Advanced\Result->execute() called at [vendor/magento/framework/Interception/Interceptor.php:58]
#5 Magento\CatalogSearch\Controller\Advanced\Result\Interceptor->___callParent() called at [vendor/magento/framework/Interception/Interceptor.php:138]
#6 Magento\CatalogSearch\Controller\Advanced\Result\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/framework/App/Action/Plugin/ActionFlagNoDispatchPlugin.php:51]
#7 Magento\Framework\App\Action\Plugin\ActionFlagNoDispatchPlugin->aroundExecute() called at [vendor/magento/framework/Interception/Interceptor.php:135]
#8 Magento\CatalogSearch\Controller\Advanced\Result\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/framework/Interception/Interceptor.php:153]
#9 Magento\CatalogSearch\Controller\Advanced\Result\Interceptor->___callPlugins() called at [generated/code/Magento/CatalogSearch/Controller/Advanced/Result/Interceptor.php:23]
#10 Magento\CatalogSearch\Controller\Advanced\Result\Interceptor->execute() called at [vendor/magento/framework/App/Action/Action.php:111]
#11 Magento\Framework\App\Action\Action->dispatch() called at [vendor/magento/framework/Interception/Interceptor.php:58]
#12 Magento\CatalogSearch\Controller\Advanced\Result\Interceptor->___callParent() called at [vendor/magento/framework/Interception/Interceptor.php:138]
#13 Magento\CatalogSearch\Controller\Advanced\Result\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/framework/Interception/Interceptor.php:153]
#14 Magento\CatalogSearch\Controller\Advanced\Result\Interceptor->___callPlugins() called at [generated/code/Magento/CatalogSearch/Controller/Advanced/Result/Interceptor.php:32]
#15 Magento\CatalogSearch\Controller\Advanced\Result\Interceptor->dispatch() called at [vendor/magento/framework/App/FrontController.php:186]
#16 Magento\Framework\App\FrontController->processRequest() called at [vendor/magento/framework/App/FrontController.php:118]
#17 Magento\Framework\App\FrontController->dispatch() called at [vendor/magento/framework/Interception/Interceptor.php:58]
#18 Magento\Framework\App\FrontController\Interceptor->___callParent() called at [vendor/magento/framework/Interception/Interceptor.php:138]
#19 Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/module-store/App/FrontController/Plugin/RequestPreprocessor.php:99]
#20 Magento\Store\App\FrontController\Plugin\RequestPreprocessor->aroundDispatch() called at [vendor/magento/framework/Interception/Interceptor.php:135]
#21 Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/module-page-cache/Model/App/FrontController/BuiltinPlugin.php:71]
#22 Magento\PageCache\Model\App\FrontController\BuiltinPlugin->aroundDispatch() called at [vendor/magento/framework/Interception/Interceptor.php:135]
#23 Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/framework/Interception/Interceptor.php:153]
#24 Magento\Framework\App\FrontController\Interceptor->___callPlugins() called at [generated/code/Magento/Framework/App/FrontController/Interceptor.php:23]
#25 Magento\Framework\App\FrontController\Interceptor->dispatch() called at [vendor/magento/framework/App/Http.php:116]
#26 Magento\Framework\App\Http->launch() called at [vendor/magento/framework/App/Bootstrap.php:263]
#27 Magento\Framework\App\Bootstrap->run() called at [pub/index.php:29]
```

**Issue 2:** strpos() expects parameter 1 to be string, array given | magento/module-eav #33588
**Steps to reproduce:**
1. Admin: Stores -> Attributes -> Product - > Add New Attribute /Choose existing attribute ex: color
Input Type should be: Dropdown
Storefront Properties: Use in Search: Yes, Visible in Advanced Search: Yes
2. Front-end - Go to http://yourmagentostore.com/catalogsearch/advanced/result/?attributename[][error]=1
Ex: https://magento2.4.local/catalogsearch/advanced/result/?test_data[][error]=1
 **Issue: There has been an error processing your request Error message is getting displayed
    Note: There is no issue with Text Attribute. URL is working for Text attribute. Issue is observed for dropdown attribute.**

Magentoroot/var/log/exception.log:
```
Exception #0 (Exception): Warning: strpos() expects parameter 1 to be string, array given in vendor/magento/module-eav/Model/Entity/Attribute/Source/Table.php on line 152

#1 strpos() called at [vendor/magento/module-eav/Model/Entity/Attribute/Source/Table.php:152]
#2 Magento\Eav\Model\Entity\Attribute\Source\Table->getOptionText() called at [vendor/magento/module-catalog-search/Model/Advanced.php:404]
#3 Magento\CatalogSearch\Model\Advanced->getPreparedSearchCriteria() called at [vendor/magento/module-catalog-search/Model/Advanced.php:206]
#4 Magento\CatalogSearch\Model\Advanced->addFilters() called at [vendor/magento/module-catalog-search/Controller/Advanced/Result.php:62]
#5 Magento\CatalogSearch\Controller\Advanced\Result->execute() called at [vendor/magento/framework/Interception/Interceptor.php:58]
#6 Magento\CatalogSearch\Controller\Advanced\Result\Interceptor->___callParent() called at [vendor/magento/framework/Interception/Interceptor.php:138]
#7 Magento\CatalogSearch\Controller\Advanced\Result\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/framework/App/Action/Plugin/ActionFlagNoDispatchPlugin.php:51]
#8 Magento\Framework\App\Action\Plugin\ActionFlagNoDispatchPlugin->aroundExecute() called at [vendor/magento/framework/Interception/Interceptor.php:135]
#9 Magento\CatalogSearch\Controller\Advanced\Result\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/framework/Interception/Interceptor.php:153]
#10 Magento\CatalogSearch\Controller\Advanced\Result\Interceptor->___callPlugins() called at [generated/code/Magento/CatalogSearch/Controller/Advanced/Result/Interceptor.php:23]
#11 Magento\CatalogSearch\Controller\Advanced\Result\Interceptor->execute() called at [vendor/magento/framework/App/Action/Action.php:111]
#12 Magento\Framework\App\Action\Action->dispatch() called at [vendor/magento/framework/Interception/Interceptor.php:58]
#13 Magento\CatalogSearch\Controller\Advanced\Result\Interceptor->___callParent() called at [vendor/magento/framework/Interception/Interceptor.php:138]
#14 Magento\CatalogSearch\Controller\Advanced\Result\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/framework/Interception/Interceptor.php:153]
#15 Magento\CatalogSearch\Controller\Advanced\Result\Interceptor->___callPlugins() called at [generated/code/Magento/CatalogSearch/Controller/Advanced/Result/Interceptor.php:32]
#16 Magento\CatalogSearch\Controller\Advanced\Result\Interceptor->dispatch() called at [vendor/magento/framework/App/FrontController.php:186]
#17 Magento\Framework\App\FrontController->processRequest() called at [vendor/magento/framework/App/FrontController.php:118]
#18 Magento\Framework\App\FrontController->dispatch() called at [vendor/magento/framework/Interception/Interceptor.php:58]
#19 Magento\Framework\App\FrontController\Interceptor->___callParent() called at [vendor/magento/framework/Interception/Interceptor.php:138]
#20 Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/module-store/App/FrontController/Plugin/RequestPreprocessor.php:99]
#21 Magento\Store\App\FrontController\Plugin\RequestPreprocessor->aroundDispatch() called at [vendor/magento/framework/Interception/Interceptor.php:135]
#22 Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/module-page-cache/Model/App/FrontController/BuiltinPlugin.php:71]
#23 Magento\PageCache\Model\App\FrontController\BuiltinPlugin->aroundDispatch() called at [vendor/magento/framework/Interception/Interceptor.php:135]
#24 Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}() called at [vendor/magento/framework/Interception/Interceptor.php:153]
#25 Magento\Framework\App\FrontController\Interceptor->___callPlugins() called at [generated/code/Magento/Framework/App/FrontController/Interceptor.php:23]
#26 Magento\Framework\App\FrontController\Interceptor->dispatch() called at [vendor/magento/framework/App/Http.php:116]
#27 Magento\Framework\App\Http->launch() called at [vendor/magento/framework/App/Bootstrap.php:263]
#28 Magento\Framework\App\Bootstrap->run() called at [pub/index.php:29]
```
---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [X] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by Morgy93** [comment]
@magento give me 2.4-develop instance

**Comment by Morgy93** [comment]
@magento give me 2.4-develop instance

**Comment by Morgy93** [comment]
Verified the issue on Magento 2.4-develop branch and the issue is reproducible:
Steps to reproduce:
1. Go to `domain.tld/catalogsearch/advanced/result/?price[from][]=1&price[to]=1` (e.g. https://52a58bd3ae67a73e5814b62433f13aee-2-4-develop.instances.magento-community.engineering/catalogsearch/advanced/result/?price[from][]=1&price[to]=1)
2. See error
![image](https://user-images.githubusercontent.com/7961978/126982506-0a7f304f-082d-451b-8be2-b22a9f9908f7.png)

**Comment by engcom-November** [comment]
Verified the issue on Magento 2.4-develop branch and the issue is reproducible:
On trying to access below URL from front-end, "There has been an error processing your request" error message is getting displayed on UI:
http://magento2.4.local/catalogsearch/advanced/result/?price[from][]=1&price[to]=1
![image](https://user-images.githubusercontent.com/60198231/127472903-9231c34a-7675-401c-af28-fcea4ce40150.png)

Magentoroot/var/log/exception.log:

main.CRITICAL: Warning: strlen() expects parameter 1 to be string, array given in /var/www/html/magento2.4/magento2/app/code/Magento/CatalogSearch/Model/Advanced.php on line 380 {"report_id":"79cdf9e388c36ce3b8cc8d55f705fca8f8e6a566f6b8c18c11190cc9c4eb90d1","exception":"[object] (Exception(code: 0): Warning: strlen() expects parameter 1 to be string, array given in /var/www/html/magento2.4/magento2/app/code/Magento/CatalogSearch/Model/Advanced.php on line 380 at /var/www/html/magento2.4/magento2/lib/internal/Magento/Framework/App/ErrorHandler.php:61)"} []

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-November 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-42934` were created

**Issue Available**: @engcom-November, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by SilinMykola** [comment]
@magento I am working on this

**Comment by andrewbess** [comment]
Hello @Morgy93 
I am not sure issue is correct.
In basically, you have wrong search request there.
So, this issue is wrong (IMHO)

cc @engcom-November cc @SilinMykola

**Comment by Morgy93** [comment]
> Hello @Morgy93
> I am not sure issue is correct.
> In basically, you have wrong search request there.
> So, this issue is wrong (IMHO)
> 
> cc @engcom-November cc @SilinMykola

It triggers a 500 server error so it is an issue, isn't it? =)

It was discovered while someone tried to hack into the system, so yes, it might not be "usual customer behavior", but still an (security) issue.

**Comment by engcom-November** [comment]
@andrewbess for wrong Search requests like this, User should not be displayed with exception/unhandled errors. Error handling should be made for better user experience and to avoid security issues. 
@Morgy93, Similar kind of [issue](https://github.com/magento/magento2/issues/33588) but has different error messages is closed as duplicate and updated here in the description.
Thank you

**Comment by andrewbess** [comment]
I understand you @engcom-November @Morgy93 
But I think we shouldn't process all request with wrong GET params.
Of course, end user shouldn't see report message of the 500 error.
I think it would be better to redirect user to page 404 in this case, but don't process wrong request.
Anyway we should consult behaviour with product owners if we want to process it.

cc @hostep @eduard13

**Comment by hostep** [comment]
I think the validation of those parameters should be broadened a bit, currently it only [validates if `from` or `to` aren't empty](https://github.com/magento/magento2/blob/2.4.2-p1/app/code/Magento/CatalogSearch/Model/Advanced.php#L368) (whatever that means, I really hate the `empty` function in php btw).
So we should probably expand the validation to check that those values match what is expected, and if those don't validate , then return an empty string or `false` (whatever makes sense).

Just my 2 cents I'm not very familiar with this part in magento's codebase.

**Comment by sidolov** [comment]
Hi @Morgy93. Thank you for your report. 
The issue has been fixed in magento/magento2#33682 by @SilinMykola in 2.4-develop branch
Related commit(s):
 - [58adfbf9c846014403f2f37d75ffab02f1b38662](https://github.com/magento/magento2/commit/58adfbf9c846014403f2f37d75ffab02f1b38662)
 - [0c72905cdfeb5181a81470002a16d1b50ba36a28](https://github.com/magento/magento2/commit/0c72905cdfeb5181a81470002a16d1b50ba36a28)
 - [d25aba755c15eb5b0337a2e7de12b79e7eaad0d3](https://github.com/magento/magento2/commit/d25aba755c15eb5b0337a2e7de12b79e7eaad0d3)
 - [e5df80f17d0457d4a95af4fa30caefd33163c013](https://github.com/magento/magento2/commit/e5df80f17d0457d4a95af4fa30caefd33163c013)
 - [dd0124b7ff79779280edeb0ed70febca6bbbccad](https://github.com/magento/magento2/commit/dd0124b7ff79779280edeb0ed70febca6bbbccad)
 - [44ed899f0925aecf725d777c889d1b5fd0d29e1c](https://github.com/magento/magento2/commit/44ed899f0925aecf725d777c889d1b5fd0d29e1c)

The fix will be available with the upcoming 2.4.3 release.
